ambry-store/src/main/java/com.github.ambry.store/DiskManager.java

```diff
@@ -86,41 +94,52 @@
-            @Override
-            public void run() {
-              try {
-                partitionAndStore.getValue().start();
-              } catch (Exception e) {
-                numFailures.incrementAndGet();
-                metrics.totalStoreStartFailures.inc();
-                logger.error("Exception while starting store for the partition" + partitionAndStore.getKey(), e);
-              }
-            }
-          }, false);
-          thread.start();
-          startupThreads.add(thread);
-        }
-        for (Thread startupThread : startupThreads) {
-          startupThread.join();
-        }
-        if (numFailures.get() > 0) {
-          logger.error(
-              "Could not start " + numFailures.get() + " out of " + stores.size() + " stores on the disk " + disk);
+      checkMountPathAccessible();
+
+      List<Thread> startupThreads = new ArrayList<>();
+      for (final Map.Entry<PartitionId, BlobStore> partitionAndStore : stores.entrySet()) {
+        Thread thread = Utils.newThread("store-startup-" + partitionAndStore.getKey(), () -> {
+          try {
+            partitionAndStore.getValue().start();
+          } catch (Exception e) {
+            numFailures.incrementAndGet();
+            logger.error("Exception while starting store for the partition" + partitionAndStore.getKey(), e);
+          }
+        }, false);
+        thread.start();
+        startupThreads.add(thread);
+      }
+      for (Thread startupThread : startupThreads) {
+        startupThread.join();
+      }
+      if (numFailures.get() > 0) {
+        logger.error(
+            "Could not start " + numFailures.get() + " out of " + stores.size() + " stores on the disk " + disk);
+      }
+
+      // DiskSpaceAllocator startup. This happens after BlobStore startup because it needs disk space requirements
+      // from each store.
+      List<DiskSpaceRequirements> requirementsList = new ArrayList<>();
+      for (BlobStore blobStore : stores.values()) {
+        if (blobStore.isStarted()) {
+          DiskSpaceRequirements requirements = blobStore.getDiskSpaceRequirements();
+          if (requirements != null) {
+            requirementsList.add(requirements);
+          }
         }
-        compactionManager.enable();
-      } else {
-        metrics.diskMountPathFailures.inc();
-        metrics.totalStoreStartFailures.inc(stores.size());
-        logger.error("Mount path does not exist: " + mountPath + " ; cannot start stores on this disk");
       }
+      diskSpaceAllocator.initializePool(requirementsList);
+
+      compactionManager.enable();
+
+      running = true;
+    } catch (StoreException e) {
+      logger.error("Error while starting the DiskManager for " + disk.getMountPath()
+          + " ; no stores will be accessible on this disk.", e);
     } finally {
+      metrics.totalStoreStartFailures.inc(!running ? stores.size() : numFailures.get());
```

Comment:
If running is false that indicates that all stores on the disk are unusable (because of diskspaceallocator init failure).